zebra-chain/src/transaction/hash.rs

```diff
@@ -19,11 +19,10 @@ pub struct Hash(pub [u8; 32]);
 
 impl<'a> From<&'a Transaction> for Hash {
     fn from(transaction: &'a Transaction) -> Self {
-        let mut hash_writer = sha256d::Writer::default();
-        transaction
-            .zcash_serialize(&mut hash_writer)
-            .expect("Transactions must serialize into the hash.");
-        Self(hash_writer.finish())
+        let hasher = TxIdHasher::new(&transaction);
+        let txid = hasher.txid();
+        // XXX: avoid unwrap here?
+        txid.unwrap()
```

Comment:
Actually, I think a panic is ok here.

In general, Zebra panics when we have an internal bug and can't recover or continue.

`read` returns `io::Error` if the data is the wrong size or structurally invalid. But that should be impossible if Zebra's structural validation matches `librustzcash`.

If there is a mismatch, then a panic will stop us creating transactions based on invalid data.

```suggestion
       hasher.txid().expect("zcash_primitives and Zebra transaction formats must be compatible")
```